From 393a32128a509670682af987bc0f349a1a66c388 Mon Sep 17 00:00:00 2001
From: Arulanbu Balusamy <quic_abalusam@quicinc.com>
Date: Mon, 13 Nov 2023 14:35:02 +0530
Subject: [PATCH] wifi: mac80211: Add the new flag for Green AP power save mode.

Resolve the "ieee80211_link_info_change_notify" calltrace issue in AP
during power saving mode.The calltrace triggered due to the WARN_ON().
From the Green AP change the BSS_CHANGED_PS flag is shared for both AP
and STA. This cause the WARN_ON during Green AP testing.

So in this change introduce the seperate ps flag "BSS_CHANGED_AP_PS"
only for AP.

Also revert the changes in "ieee80211_recalc_ps_vif" function. Because
this function is supported only for STA mode and it is a vif specific notifier.

Updated flags:
================================
BSS_CHANGED_PS     - vif type STA
BSS_CHANGED_AP_PS  - vif type AP

Signed-off-by: Arulanbu Balusamy <quic_abalusam@quicinc.com>
---
 include/net/mac80211.h | 65 +++++++++++++++++++++---------------------
 net/mac80211/cfg.c     |  6 ++--
 net/mac80211/mlme.c    |  4 +--
 3 files changed, 38 insertions(+), 37 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -319,7 +319,7 @@
  * @BSS_CHANGED_IDLE: Idle changed for this BSS/interface.
  * @BSS_CHANGED_SSID: SSID changed for this BSS (AP and IBSS mode)
  * @BSS_CHANGED_AP_PROBE_RESP: Probe Response changed for this BSS (AP mode)
- * @BSS_CHANGED_PS: PS changed for this BSS (both AP and STA mode)
+ * @BSS_CHANGED_PS: PS changed for this BSS (STA mode)
  * @BSS_CHANGED_TXPOWER: TX power setting changed for this interface
  * @BSS_CHANGED_P2P_PS: P2P powersave settings (CTWindow, opportunistic PS)
  *	changed
@@ -341,6 +341,8 @@
  * @BSS_CHANGED_FILS_DISCOVERY: FILS discovery status changed.
  * @BSS_CHANGED_UNSOL_BCAST_PROBE_RESP: Unsolicited broadcast probe response
  *	status changed.
+ * @BSS_CHANGED_AP_PS: PS changed for this BSS (AP mode)
+ *
  *
  */
 enum ieee80211_bss_change {
@@ -376,6 +378,7 @@
 	BSS_CHANGED_HE_BSS_COLOR	= 1<<29,
 	BSS_CHANGED_FILS_DISCOVERY      = 1<<30,
 	BSS_CHANGED_UNSOL_BCAST_PROBE_RESP = 1<<31,
+	BSS_CHANGED_AP_PS               = BIT_ULL(32),
 
 	/* when adding here, make sure to change ieee80211_reconfig */
 };
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1660,9 +1660,9 @@
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
 	struct ieee80211_link_data *link;
 	struct beacon_data *old;
-	int err = -EINVAL;
+	s64 err = -EINVAL;
 	struct ieee80211_bss_conf *link_conf;
-	u32 changed = 0;
+	u64 changed = 0;
 
 	sdata_assert_lock(sdata);
 
@@ -1692,9 +1692,9 @@
 	    params->ap_ps_valid) {
 		link_conf->ap_ps_enable = params->ap_ps_enable;
 		if (err < 0)
-			err = BSS_CHANGED_PS;
+			err = BSS_CHANGED_AP_PS;
 		else
-			err |= BSS_CHANGED_PS;
+			err |= BSS_CHANGED_AP_PS;
 	}
 
 	if (params->beacon.he_bss_color_valid &&
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -2261,25 +2261,11 @@
 
 void ieee80211_recalc_ps_vif(struct ieee80211_sub_if_data *sdata)
 {
-	struct ieee80211_link_data *link;
-	u16 link_id;
 	bool ps_allowed = ieee80211_powersave_allowed(sdata);
 
 	if (sdata->vif.cfg.ps != ps_allowed) {
 		sdata->vif.cfg.ps = ps_allowed;
-		if (!sdata->vif.valid_links) {
-			ieee80211_link_info_change_notify(sdata, &sdata->deflink,
-							  BSS_CHANGED_PS);
-		} else {
-			for_each_set_bit(link_id, &sdata->vif.valid_links,
-					 IEEE80211_MLD_MAX_NUM_LINKS) {
-				link = sdata_dereference(sdata->link[link_id], sdata);
-				if (!link)
-					continue;
-
-				ieee80211_link_info_change_notify(sdata, link, BSS_CHANGED_PS);
-			}
-		}
+		ieee80211_vif_cfg_change_notify(sdata, BSS_CHANGED_PS);
 	}
 }
 
